api/server/sdk/watcher.go

```diff
@@ -0,0 +1,242 @@
+)
+
+type WatcherServer struct {
+	volumeServer *VolumeServer
+
+	watchConnections map[string][]*watchConnection
+	volumeChannel    chan *api.Volume
+	sync.RWMutex
+}
+
+// Watch streams a list of events back to clients. Events are resources that can typically fetch with openstorage API, such as Volumes,
+// Nodes, Disks, etc.
+//
+// Implementation note: the flow of data starts from porx ETCD changes, then the data will be send to openstorage via a golang channel.
+// Once the event object arrives at openstorage, it will be redistributed to a list of watch connections via another set of channels.
+func (w *WatcherServer) Watch(req *api.SdkWatchRequest, stream api.OpenStorageWatch_WatchServer) error {
+	if req.GetVolumeEvent() != nil {
+		return w.volumeWatch(req.GetVolumeEvent(), stream)
+	}
+	return nil
+}
+
+type watchConnection struct {
+	name         string
+	eventType    string
+	eventChannel chan interface{}
+}
+
+func (w *watchConnection) callBack(eventData interface{}) {
+	select {
+	case w.eventChannel <- eventData:
+		logrus.Debugf("successfully callback event for %v", w.name)
+	default:
+		logrus.Warnf("failed to send eventData for %v with event type %v", w.name, w.eventType)
+	}
+
+}
+
+func (s *WatcherServer) registerWatcher(client *watchConnection, eventType string) {
+	s.Lock()
+	defer s.Unlock()
+	if s.watchConnections == nil {
+		s.watchConnections = make(map[string][]*watchConnection)
+	}
+	s.watchConnections[eventType] = append(s.watchConnections[eventType], client)
+	logrus.Debugf("successfully register watcher %v", client.name)
+}
+
+func (s *WatcherServer) removeWatcher(name string, eventType string) {
+	s.Lock()
+	defer s.Unlock()
+	var newWatchers []*watchConnection
+	for _, client := range s.watchConnections[eventType] {
+		if client.name == name {
+			// clean up client go channel
+			close(client.eventChannel)
+			continue
+		}
+		newWatchers = append(newWatchers, client)
+	}
+	s.watchConnections[eventType] = newWatchers
+}
+
+func (s *WatcherServer) startWatcher(ctx context.Context, done chan bool) error {
+	group, _ := errgroup.WithContext(ctx)
+	errChan := make(chan error)
+	group.Go(func() error {
+		return s.startVolumeWatcher(ctx, done)
+	})
+
+	// wait for err-group processes to be done
+	go func() {
+		errChan <- group.Wait()
+	}()
+
+	// wait only as long as context deadline allows
+	select {
+	case err := <-errChan:
+		if err != nil {
+			return status.Errorf(codes.Internal, "error starting watcher: %v", err)
+		} else {
+			return nil
+		}
+	case <-ctx.Done():
+		return status.Error(codes.DeadlineExceeded,
+			"Deadline is reached, server side func exiting")
+	}
+}
+
+func (s *WatcherServer) startVolumeWatcher(ctx context.Context, done chan bool) error {
+	if s.watchConnections == nil {
+		s.watchConnections = make(map[string][]*watchConnection)
+	}
+
+	// wait for driver to be initialized with an non-empty volume watcher
+	for {
+		if s.volumeServer.driver(ctx) == nil {
+			continue
```

Comment:
This will spin continuously if invoked before driver is initialized. Add a sleep may be?

Reply:
Please add a sleep here before merging the PR